### ISSUE TITLE
Bidi text edit support

### ISSUE THREAD

**Issue by ebraminio** [microsoft/vscode#65648]
#11770 is so vague that I can't even understand what is the requested feature... a clear and achievable change to VSCode for RTL is I believe starts from this, please make a paragraph RTL if it starts with RTL text.

Steps to reproduce:
Put this on gedit:
```
متن
متن
متن
متن
text
text
text
text
```
![image](https://user-images.githubusercontent.com/833473/50404768-23125f80-07c1-11e9-973e-7a477690be5b.png)

Now try the same with VSCode

![image](https://user-images.githubusercontent.com/833473/50404780-44734b80-07c1-11e9-92b1-04a457bf90a0.png)

Expected:
Easily achievable by adding `dir=auto` to div element and remove `dir=ltr` from the span, I just did it for the first line

![image](https://user-images.githubusercontent.com/833473/50404833-da0edb00-07c1-11e9-9294-86bda463d797.png)

Alternatively, you can put a switch (toggle-able with command switch, something like > Toggle Direction) and apply document direction switch to whole document and detect whether a document needs RTL direction behind the scene, if you don't like per paragraph direction setting.

**Comment by ebraminio** [comment]
@vscodebot not related

**Comment by ebraminio** [comment]
Thanks for assigning this.

What I am requesting is entering this code,

 `$$('.mtk1[dir=ltr]').forEach(x => { x.removeAttribute('dir'); x.parentElement.parentElement.setAttribute('dir', 'auto'); })`

Which turns

![image](https://user-images.githubusercontent.com/833473/50425308-c1b5c380-0888-11e9-8b53-6d1b2b12a44e.png)

Into this,

![image](https://user-images.githubusercontent.com/833473/50425311-cda18580-0888-11e9-92a2-f1c9ede24650.png)

The hard part however is to fix the issues afterward like incorrect assumption of text place,

![image](https://user-images.githubusercontent.com/833473/50425312-ec078100-0888-11e9-88e9-c8f4651c9ae5.png)

So maybe you should do the direction detection in JavaScript so you will be in control of it and know what is its value.

**Comment by ebraminio** [comment]
> So maybe you should do the direction detection in JavaScript so you will be in control of it and know what is its value.

You can also use dir=auto but detect the assigned direction using getComputedStyle I guess.

**Comment by michael-james** [comment]
Thank you for working on this! I would like my department at work (we correct translated HTML and work with Arabic and Hebrew) to start using VS Code but I know that the ability to view these RTL languages right-align will be a crucial requirement.

**Comment by sarmadka** [comment]
What's the status of this? I am also looking for this solution, and it looks like many people are looking for it as well.

**Comment by lwcvl** [comment]
I strongly support this request. Connected to this issue, also consider how punctuation behaves in this matter. Right now, VS Code will place a period at the end of a sentence to the very right of the Arabic text, like: 

شيء مكتوب في اللغة العربية.  (WRONG)

Visibly different from:

.شيء مكتوب في اللغة العربية (CORRECT)

This can also mess up brackets, making them face the wrong way:

)٢٣(   (WRONG)

Instead of:

(٢٣) (CORRECT)

**Comment by drewpotter** [comment]
👍 Please fix RTL layout! 🙏 

I am having the mentioned issue with brackets...

<img width="397" alt="Screenshot 2019-08-10 at 13 52 14" src="https://user-images.githubusercontent.com/7907456/62822057-1850e300-bb76-11e9-9ba0-8f74b0c20050.png">

**Comment by ebraminio** [comment]
Thanks for the report. Let's not confuse the issues which is the case for #11770, this one should be fixed with #78867 to avoiding confusion. Thanks

**Comment by zspitz** [comment]
@ebraminio I think the brackets problem arises because the paragraph direction is always set to LTR, as I've described [here](https://github.com/microsoft/vscode/issues/78867#issuecomment-581910560). Fixing the paragraph direction should fix the brackets issue.

**Comment by ebraminio** [comment]
Even the fact paragraph direction solves the issue for some of the cases but it doesn't always solve the issue where paragraph or document direction is not the same as the text, like the two above comments, bidi bracket is a separate matter http://www.unicode.org/notes/tn39/ which deserves to get fixed separately, let's don't mix that to this where is possible to keep them separate, mixing them makes just harder to vscode devs to know what is needed to get fixed.

**Comment by zspitz** [comment]
@ebraminio Agreed.

Which comments are you referring to? [This comment](https://github.com/microsoft/vscode/issues/65648#issuecomment-520146687) is independent of paragraph direction, but AFAICT [this comment](https://github.com/microsoft/vscode/issues/65648#issuecomment-484588643) happens because of paragraph direction -- AFAICT a period within an Arabic text run should render on the left, instead of on the right. @lwcvl Can you confirm this?

**Comment by Drkingman** [comment]
درود

**Comment by remcohaszing** [comment]
I’m going to have a look at this. I don’t have any prior experience with RTL languages, so please bear with me and correct me if I’m wrong.

As far as I understand, a paragraph can be LTR or RTL. If the paragraph is LTR, the content should be aligned to the left. If the paragraph is RTL, the content should be aligned to the right. When an arrow key is pressed on the keyboard, the cursor should move into the direction the arrow key points. If a paragraph contains text of the opposite direction, horizontal cursor movement should be inverted.

VSCode actually treats RTL text pretty well, but it’s forced into LTR mode. This means that the content is always aligned to the left. And using arrow keys to move the cursor work in the opposite direction.

VSCode works with plain text. In plain text there is no concept of neither paragraphs nor text direction. There is a concept of lines, which is similar to a paragraph. For example, let’s take the example in the OP:

```
متن
متن
متن
متن
text
text
text
text
```

This can be interpreted as markdown. In that case it’s equivalent to:

```
متن متن متن متن text text text text
```

Funnily GitHub considers the same code block as RTL if the language is set to `md`

```md
متن متن متن متن text text text text
```

or to the following HTML:

```html
<p>متن متن متن متن text text text text</p>
```

There are no clear rules how an editor should render it. From some quick testing I see gedit detects the direction based on the first character on the line. This isn’t objectively right or wrong.

VSCode / Monaco Editor has a concept of decorations. Decorations can be used to make text bold, italic, inject spacing, give it a background color, increase line height, etc. Decorations can also apply to a whole line.

My plan is to allow whole line decorations to determine text direction of that line. Decorations can overlap. If there are more RTL decorations than LTR decorations on that line, the line is RTL. If there are more or an equal amount of LTR decorations, LTR wins, making it the default.

This will allow users of Monaco editor or VSCode extension developers to implement RTL as they see fit. For example an extension could base it on the first character, a count of LTR vs RTL characters, manual key bindings, etc.

Removing forced text direction Monaco adds, shows that this largely already works. There are still some things to resolve however. Some quirks I found are:

- The cursor direction is currently forced LTR.
- Some things are always rendered on the left. For example if a text selection contains the first character of a line.
- The scrollbar shouldn’t overlap with RTL text.

**Comment by ebraminio** [comment]
> When an arrow key is pressed on the keyboard, the cursor should move into the direction the arrow key points. If a paragraph contains text of the opposite direction, horizontal cursor movement should be inverted.

Some editors follow visual editor for normal cursor movement but while selection go for logical cursor movement.

> VSCode works with plain text. In plain text there is no concept of neither paragraphs nor text direction.

True and for example Firefox goes with auto detection of direction, `data:text/plain;charset=utf8,متن` Some say other browsers are "being silly" to not do that https://github.com/w3c/alreq/issues/257#issuecomment-1238320892 and I agree that's better than having no other way for that.

<img width="569" alt="Image" src="https://github.com/user-attachments/assets/51f08309-1386-4ddc-afde-646fc837745b" />

> There are no clear rules how an editor should render it. From some quick testing I see gedit detects the direction based on the first character on the line. This isn’t objectively right or wrong.

True. I like the approach you are taking, the current situation is there is no way to edit in RTL paragraphs and that's so intolerable that's why the 'gedit' like approach is at least better than the current situation. Ideally there should be some way to instruct the editor the following plain text is RTL (as you say maybe via an extension) but in a plaintext some direction detection mechanism also can be helpful, it will be nice to have something like tab size configuration in vscode for direction and having a dir="auto" as the default.

**Comment by m2-farzan** [comment]
@remcohaszing 

> There are no clear rules how an editor should render it. From some quick testing I see gedit detects the direction based on the first character on the line. This isn’t objectively right or wrong.

Hi! I'm an RTL-speaking user. Apologies for not having a lot of context about how VSCode works behind the scenes. I want to offer some feedback, highlight some missing points, and generally dump info. As a side note, I have implemented RTL support for jupyter notebook ([link](https://github.com/jupyter/notebook/pull/5052#issuecomment-647725947)), but I'm not about to suggest whether or not you should take the same approach I took there.

First, as you pointed out, each unicode character is either LTR, RTL, or Neutral. As a user, the vast majority of the apps I deal with use the first character of a paragraph to determine directionality. Examples include gedit, google docs, html elements with `dir=auto` in all major browsers, and most of the input boxes in mobile apps. Even this github input box that I'm typing in starts to behave like that if I add a `dir="auto"` attribute to it.

Subjectively, it's a little unexpected when I come across apps that use other heuristics. IIRC, google docs (maybe twitter too) used to be one of those, checking the total number of RTL chars against LTR ones to determine direction of a paragraph. Now, that makes some sense logically, but it felt unfamiliar and also the line jumped around and was hard to control, especially at the beginning of the sentence!

On the other hand, the big flaw of using the first character still exists. What if one is writing an RTL paragraph but the first word is needed to be written in LTR? This makes the whole sentence look LTR in editor. As a user, there are three solutions I usually consider:
1. An unpleasant but quick solution is to rephrase my sentence or add some filler adverb such that the sentence starts with an RTL word again. Not acceptable but practically my first thought (I'm lazy), so worth mentioning!
2. If it's a rich text editor like google docs and microsoft word, then I can simply override the automatic direction detection by clicking some button, so it's not an issue with those. This kind of annotation is usually assigned to the whole paragraph, not to some part of the text within it, though MS word has a lesser-known concept of RTL-run that can be assigned a span of text within a paragraph too. Also, a weird thing I remember is MS notepad having a "make paragraph RTL" option when you right-clicked on a paragraph, even though it wasn't a rich text editor. To this day idk how it works, but I don't have windows now so I can't even verify if it's still there. I guess it may have been some in-memory hack that got lost after you close the document. All in all, this approach doesn't go a long way with plain text, right? My latex source file doesn't have a way to store this information, so all these annotations will be lost as soon as I close it.
3. Finally, the best solution is to use an "RTL marker". It's just an invisible RTL character that is specifically designed for this purpose in unicode. I can put it at the beginning of the sentence, and it fixes all the problems in my life. This is the most robust and flexible solution, and is recommended in a w3 [article](https://www.w3.org/International/questions/qa-bidi-unicode-controls) (good read in general). The downsides are: (1) It's not common knowledge at all, even among most RTL writers. (2) There is no key combination for it on the keyboard, unless I define a custom one and (3) some apps (like twitter IIRC) may randomly decide to remove it from the text as a part of their "sanitization". Also note that this approach solves some ambiguities that are too niche to be resolvable by a paragraph direction annotation described in (2) -- see the article linked.


> When an arrow key is pressed on the keyboard, the cursor should move into the direction the arrow key points. If a paragraph contains text of the opposite direction, horizontal cursor movement should be inverted.

This is not necessarily true. Google docs simply flips the logical direction, which means the direction will be reversed when navigating within an LTR segment in RTL text or vice versa. IIRC, microsoft word does the same too.

**Comment by remcohaszing** [comment]
That’s all interesting feedback. Thanks!

I didn’t know about the [RTL](https://unicode-explorer.com/c/200F) marker. There’s also an [LTR](https://unicode-explorer.com/c/200E) marker. I see there’s also [ALM](https://unicode-explorer.com/c/061C). I should have a closer look at those.

I’m also curious to get input from @aiday-mar about my suggested approach.

**Comment by aiday-mar** [comment]
Hi @remcohaszing thank you and everyone else for the comments. I must admit I am not very familiar with this feature area, but I will talk to my team about this ask and the proposed solutions.

**Comment by mehdisadeghi** [comment]
> When an arrow key is pressed on the keyboard, the cursor should move into the direction the arrow key points. If a paragraph contains text of the opposite direction, horizontal cursor movement should be inverted.

Just for clarity, it should also include character and word selection, as well as selection inversion–i.e., when pressing an arrow key in combination with the Control, Shift, or both keys. This aspect is often overlooked.

Edit: The same applies when replacing the arrow keys with the Home/End keys to select the entire line. Additionally, when using the the UP/Down arrow keys instead of the Left/Right arrow keys to select a line, the highlighted side should follow the same logic.

@remcohaszing

**Comment by y1n0** [comment]
> If a paragraph contains text of the opposite direction, horizontal cursor movement should be inverted.

No, it should not! it gets really confusing when the cursor moves in the opposite direction of the key, regardless of the direction of the text.

This is a clip of the current (version 1.102.0) way vscode behaves. it is so confusing to click an arrow only to have the cursor move in the opposite direction:

https://github.com/user-attachments/assets/84e68199-2353-4ac4-8bd1-d7815d64fea0

Other applications behave properly. Take Firefox address bar for example:

https://github.com/user-attachments/assets/70944f96-5c56-46a4-8118-abd761406536

And the default Apple Notes, although it still moves in the wrong direction (opposite of the key) if you hold `opt` key.

https://github.com/user-attachments/assets/431f8a73-b62f-4b33-814f-b6e77b206b0c

**Comment by y1n0** [comment]
> If a paragraph contains text of the opposite direction, horizontal cursor movement should be inverted.

No, it should not! it gets really confusing when the cursor moves in the opposite direction of the key, regardless of the direction of the text.

This is a clip of the current (version 1.102.0) way vscode behaves. it is so confusing to click an arrow only to have the cursor move in the opposite direction:

https://github.com/user-attachments/assets/1f478cac-8b71-4728-940d-60e0daebd3b6

Other applications behave properly. Take Firefox address bar for example:

https://github.com/user-attachments/assets/93f6b61f-17e5-4f5d-b24e-36a0c30ef6bd

And the default Apple Notes, although it still moves in the wrong direction (opposite of the key) if you hold `⌥` key.

https://github.com/user-attachments/assets/56b9dcad-0cbb-44e0-a830-a9b233dd2b7b

**Comment by behdad** [comment]
> > If a paragraph contains text of the opposite direction, horizontal cursor movement should be inverted.
> 
> No, it should not! it gets really confusing when the cursor moves in the opposite direction of the key, regardless of the direction of the text.

I second that.

**Comment by remcohaszing** [comment]
#255455 is merged. This adds a decoration based RTL support to Monaco Editor. It’s just not yet exposed to VSCode extensions.

**Comment by sarmadka** [comment]
> [#255455](https://github.com/microsoft/vscode/pull/255455) is merged. This adds a decoration based RTL support to Monaco Editor. It’s just not yet exposed to VSCode extensions.

THANK YOU.

Does this respect the rtl/ltr unicode markers? I looked at the code and it looks like it doesn't, but I don't know what decorations actually are so I'm not sure. If my understanding is correct, the code just counts how many characters are rtl vs ltr and decides accordingly. I think this is fine, but adding a consideration for the rtl/ltr marker would make it more solid as it would allow the user to override the default behaviour easily. I.e. it should be something like this:

```
getTextDirection:
  if first char in the line is rtl marker (U+200F)
    return RTL
  else if first char in the line is ltr marker (U+200E)
    return LTR
  else
    do the current logic
```

**Comment by remcohaszing** [comment]
Decorations are essentially a common lower-level API to attach all sorts of attributes to inline text or lines. For example, inline color pickers such as you see in CSS files and squiggly error lines are higher level APIs built on decorations. Or if you need another example, [I maintain a VSCode extension](https://github.com/remcohaszing/vscode-markdown-decorations) which applies all sorts of decorations to make markdown editing experience a bit WYSIWYG-like. Decorations can also overlap. Now such decorations may mark a line as the text direction of a line. My implementation doesn’t count any characters, it counts text direction specifying decorations on each line. If there are more RTL decorations than there are LTR decorations, the line is considered RTL.

Detection of RTL in the context of code is hard, even with explicit RTL/LTR unicode markers. Plaintext isn’t the same as a text document. Text documents have paragraphs. Plaintext has lines. Those are similar, but not the same. The meaning of plaintext depends on context. For example, the following markdown has three paragraphs:

```markdown
<p>
אני אוהב

pizza
</p>
```

But if this is interpreted as HTML, it’s just one paragraph.

I recommend the VSCode team to stay away from RTL detection. There are just too many cases, and too many opinions. Any implementation will disappoint users. It is now possible however, for third party extensions to add RTL support in various ways (assuming this API will be exposed).

**Comment by sarmadka** [comment]
> I recommend the VSCode team to stay away from RTL detection. There are just too many cases, and too many opinions. Any implementation will disappoint users. It is now possible however, for third party extensions to add RTL support in various ways (assuming this API will be exposed).

While I agree that extensions should be enabled to support the various ways, some default behaviour is necessary. A default behaviour of determining the direction from the first character will satisfy maybe 90% of the users if not more. This is what GEdit, Kate, and Gnome Builder do for example.
